frontend/include/chpl/resolution/resolution-types.h

```diff
@@ -885,6 +885,103 @@ class TypedFnSignature {
   /// \endcond DO_NOT_DOCUMENT
 };
 
+// Container for resolution candidates and (if applicable) their correponding
+// forwarding-to types.
+struct CandidatesAndForwardingInfo {
+  private:
+  std::vector<const TypedFnSignature*> candidates;
+  std::vector<types::QualifiedType> forwardingInfo;
+
+  public:
+   // Add a candidate and optional forwarding info.
+   void addCandidate(const TypedFnSignature* candidate,
+                     const types::QualifiedType* forwardingTo = nullptr) {
+     candidates.push_back(candidate);
+     if (forwardingTo) {
+       forwardingInfo.push_back(*forwardingTo);
+     }
+   }
+
+   // Compute and fill in forwarding info for a range of newly-added candidates.
+   void helpComputeForwardingTo(const CallInfo& fci, size_t start) {
+     types::QualifiedType forwardingReceiverActualType = fci.calledType();
+     forwardingInfo.resize(start);
+     for (size_t i = start; i < candidates.size(); i++) {
+       forwardingInfo.push_back(forwardingReceiverActualType);
+     }
+   }
+
+   // Move the contents of another container into this one, clearing out the
+   // other.
+   void takeFromOther(CandidatesAndForwardingInfo& other) {
+     candidates.insert(candidates.end(),
+                       std::make_move_iterator(other.candidates.begin()),
+                       std::make_move_iterator(other.candidates.end()));
+     forwardingInfo.insert(
+         forwardingInfo.end(),
+         std::make_move_iterator(other.forwardingInfo.begin()),
+         std::make_move_iterator(other.forwardingInfo.end()));
+     other.candidates.clear();
+     other.forwardingInfo.clear();
+   }
+
+   // Get the candidate at the provided index with no bounds checking.
+   const TypedFnSignature* get(size_t i) const { return candidates[i]; }
+
+   // Get the forwarding info at the provided index with no bounds checking.
+   types::QualifiedType getForwardingInfo(size_t i) const { return forwardingInfo[i]; }
+
+   // Check if any candidates are present
+   bool empty() const {
+     return candidates.empty();
+   }
+
+   // Get the number of candidates
+   size_t size() const {
+     return candidates.size();
+   }
+
+   // Return true if this container stores any forwarding info
+   bool hasForwardingInfo() const {
+     return !forwardingInfo.empty();
+   }
+
+   // Iterator over contained candidates
+   std::vector<const TypedFnSignature*>::const_iterator begin() const {
+     return candidates.begin();
+   }
+   std::vector<const TypedFnSignature*>::const_iterator end() const {
+     return candidates.end();
+   }
+
+   /* Query system supporting functions */
+
+   static bool update(CandidatesAndForwardingInfo& keep,
+                      CandidatesAndForwardingInfo& addin) {
+     return defaultUpdate(keep, addin);
+   }
+   size_t hash() const {
+     return chpl::hash(candidates, forwardingInfo);
+   }
+   void mark(Context* context) const {
+     for (const auto& candidate : candidates) {
```

Comment:
Don't we have mark functions for vectors? Since `candidate` is a raw pointer, it would match:

```C++
template<typename T> struct mark<T*> {
  void operator()(Context* context, const T* keep) const {
    context->markUnownedPointer(keep);
  }
};
```

And the vector template is as follows:

```C++
template<typename T> struct mark<std::vector<T>> {
  void operator()(Context* context, const std::vector<T>& keep) const {
    for (auto const &elt : keep) {
      chpl::mark<T> marker;
      marker(context, elt);
    }
  }
};
```

So,
```C++
marker<decltype(candidates)>{}(candidates);
marker<decltype(forwardingInfo)>{}(forwardingInfo);
```

Reply:
Ah yep, I thought there was one but couldn't find it, and saw a loop through a vector in `ResolvedExpression::mark` so figured I must have imagined it. Will change it.